cardano-crypto-praos/src/Cardano/Crypto/VRF/Praos.hs

```diff
@@ -0,0 +1,464 @@
+
+foreign import ccall "randombytes_buf" randombytes_buf :: Ptr a -> CSize -> IO ()
+
+-- Key size constants
+
+certSizeVRF :: Int
+certSizeVRF = fromIntegral crypto_vrf_proofbytes
+
+signKeySizeVRF :: Int
+signKeySizeVRF = fromIntegral crypto_vrf_secretkeybytes
+
+verKeySizeVRF :: Int
+verKeySizeVRF = fromIntegral crypto_vrf_publickeybytes
+
+vrfKeySizeVRF :: Int
+vrfKeySizeVRF = fromIntegral crypto_vrf_outputbytes
+
+-- | Allocate a 'Seed' and attach a finalizer. The allocated memory will not be initialized.
+mkSeed :: IO Seed
+mkSeed = do
+  ptr <- mallocBytes (fromIntegral crypto_vrf_seedbytes)
+  Seed <$> newForeignPtr finalizerFree ptr
+
+-- | Generate a random seed.
+-- Uses 'randombytes_buf' to create random data.
+genSeed :: IO Seed
+genSeed = do
+  seed <- mkSeed
+  withForeignPtr (unSeed seed) $ \ptr ->
+    randombytes_buf ptr crypto_vrf_seedbytes
+  return seed
+
+seedFromBytes :: ByteString -> Seed
+seedFromBytes bs | BS.length bs < fromIntegral crypto_vrf_seedbytes =
+  error "Not enough bytes for seed"
+seedFromBytes bs = unsafePerformIO $ do
+  seed <- mkSeed
+  withForeignPtr (unSeed seed) $ \ptr ->
+    BS.useAsCString bs $ \cstr ->
+      copyBytes (castPtr ptr) cstr (fromIntegral crypto_vrf_seedbytes)
+  return seed
+
+-- | Convert an opaque 'Seed' into a 'ByteString' that we can inspect. Note
+-- that this will leak the seed into unprotected memory.
+unsafeRawSeed :: Seed -> IO ByteString
+unsafeRawSeed (Seed fp) = withForeignPtr fp $ \ptr ->
+  BS.packCStringLen (castPtr ptr, fromIntegral crypto_vrf_seedbytes)
+
+-- | Convert a proof verification output hash into a 'ByteString' that we can
+-- inspect.
+outputBytes :: Output -> ByteString
+outputBytes (Output op) = unsafePerformIO $ withForeignPtr op $ \ptr ->
+  BS.packCStringLen (castPtr ptr, fromIntegral crypto_vrf_outputbytes)
+
+-- | Convert a proof into a 'ByteString' that we can inspect.
+proofBytes :: Proof -> ByteString
+proofBytes (Proof op) = unsafePerformIO $ withForeignPtr op $ \ptr ->
+  BS.packCStringLen (castPtr ptr, certSizeVRF)
+
+-- | Convert a public key into a 'ByteString' that we can inspect.
+pkBytes :: PK -> ByteString
+pkBytes (PK op) = unsafePerformIO $ withForeignPtr op $ \ptr ->
+  BS.packCStringLen (castPtr ptr, verKeySizeVRF)
+
+-- | Convert a public key into a 'ByteString' that we can inspect.
+skBytes :: SK -> ByteString
+skBytes (SK op) = unsafePerformIO $ withForeignPtr op $ \ptr ->
+  BS.packCStringLen (castPtr ptr, signKeySizeVRF)
+
+instance Show Proof where
+  show = show . proofBytes
+
+instance Eq Proof where
+  a == b = proofBytes a == proofBytes b
+
+instance ToCBOR Proof where
+  toCBOR = toCBOR . proofBytes
+  encodedSizeExpr _ _ =
+    encodedSizeExpr (\_ -> fromIntegral certSizeVRF) (Proxy :: Proxy ByteString)
+
+instance FromCBOR Proof where
+  fromCBOR = proofFromBytes <$> fromCBOR
+
+
+instance Show SK where
+  show = show . skBytes
+
+instance Eq SK where
+  a == b = skBytes a == skBytes b
```

Comment:
Do we need this? I guess it doesn't do much harm

Reply:
I think it's more harmful to not have `Eq` instances for things that are morally equality-comparable than to have them, in general. Happy to remove it though, if that's the consensus.

Reply:
@dcoutts Opinion on this? I know we've e.g. deliberately removed `Ord` instances before.